### DISCUSSION TITLE
Outer joins ?

### DISCUSSION THREAD

**Discussion by nicad** [geldata/gel#3460] [General]
Does EdgeQL support left/right outer joins ? If so can you give an example ?
Also is there a way to test if a column is missing from a result ? for example when doing polymorphic queries different types can return different columns: is there a way to test for column presence in a filter clause ?

**Comment by elprans** [comment]
> is there a way to test for column presence

Use the [exists](https://www.edgedb.com/docs/stdlib/set#operator::exists) operator.

> Does EdgeQL support left/right outer joins ? If so can you give an example ?

Link references (and computed links) in shapes are all left joins essentially, so:

```
select Movie { actors }
```

is a left join, and so is:

```
select Foo { left_joined := (select Bar filter Bar.prop = Foo.prop) }
```

Right joins are just left joins in reverse.

**Reply by jods4** [reply]
What about full outer joins?

What about inner joins? Is it a link reference combined with `exists`? Does the query compiler recognize this pattern and generate a simple inner join?

**Reply by elprans** [reply]
Well, "left join" here means a logical relational operation.  It wouldn't _necessarily_ be implemented as a SQL `LEFT JOIN`.  The queries like the shape example above are often aggregated subqueries with inner joins.

I'll probably be easier to discuss things on concrete examples.  Do you have a particular SQL query / use case in mind?

**Reply by jods4** [reply]
I use the termes inner, left, right, full, cross, anti, semi as they are defined in relational theory. 
Given collections A and B that share a relationship, I want results when both A and B match, resp. all A, etc.

An example of inner join would be "give me all users and their orders, provided they have at least one".

Left join might be "give me all users and their phone numbers, they might have none".

I rarely use right join, but it would be like starting the query on phone numbers, maybe filtering to keep only the mobile numbers, _then_ asking for the users and keeping all users.

Full outer join, I'm not sure how you'd even represent the projection in EdgeDB. Because the example I've seen so far tend to load one object with a collection inside it, which implies the "left" object always exists and that's not how full joins "work" as any side might be missing.
An example could be: I have a query that returns men, one that returns women, give me married couples (man, woman), including all single men and all single women.

Cross: I have a table of meats and a table of sauces, build a chinese menu -- all combinations ;)

Semi: give me all users that have at least one phone number (don't load the phones).

Anti: give me all users that do not have a phone number.

**Reply by jods4** [reply]
One level of complexity higher is that I might not want to load the collections in an inner/outer join, but maybe compute an aggregate,
e.g. count the orders per user.
It's easy to filter users with 0 orders (to get an "inner" effect) as long as the aggregated value can't be part of the valid set of values.

If I have the requirement "for all people that have children, what's the highest school degree of their children?", I still want the result null if someone has children that don't have a degree yet, but I don't want results for people that have no children at all. Note that the these two populations look the same in the result set.
In SQL:
```sql
select p.name, max(c.degree)
from Person p
inner join Child c on c.parent = p.id
```

If you have collections that only work in an "outer" fashion, I suppose you can add a semi-join to exclude the persons without children at all, but that's a bit less efficient as you join on the children twice:
```sql
select p.name, max(c.degree)
from Person p
left join Child c on c.parent = p.id
where exists (select * from Child where parent = p.id)
```

**Reply by elprans** [reply]
We are working on a more elaborate "EdgeQL for SQL users", but this discussion is a great opportunity to summarize that.

1. EdgeQL forgoes the explicit `JOIN` syntax (other relational query languages, like QUEL also didn't have it).
2. Outer joins are not, strictly speaking, relational operators: they are SQL extensions which naturally followed from the explicit `JOIN` syntax.  `RIGHT` and `FULL OUTER` variants are there because it makes sense for SQL syntax, not necessarily because they're needed often.
3. EdgeQL prefers and prioritizes nested structure to flat tuples.

With that, let's go through each of the join scenarios.

CROSS JOIN
==========

This is the simplest and most fundamental relational operator.  EdgeQL, like SQL and other relational query languages, computes expressions over a cross product of input relations.  Other kinds of joins are basically sugar over predicates on the result of the cross join.

So, a "chinese menu" in EdgeQL is as simple as:

```
select (Sauce.name, Meat.name)  # tuple of names
```

LEFT JOIN
=========

Object associations in query shapes are left joins, so:

```
select Meat {
  sauces := .<good_for[is Sauce]
}
```

or, if your join is over a non-key property:

```
select Meat {
  dishes := (select Dish filter .meat_name = Meat.name)
}
```

RIGHT JOIN
==========

This is just an inverse of `LEFT JOIN`.  In EdgeQL it's really easy to build and refactor queries, so, taking your phone numbers example, if you start with numbers:

```
select PhoneNumber filter .is_mobile
```

and then decide to join users, you just lift your query into a `with` alias:

```
with 
  phones := (select PhoneNumber filter .is_mobile)
select
  User {
    mobile_phones := (select phones filter .user = user)
  }
```

INNER JOIN
==========

To do this, check LEFT JOIN for existence of joined part:

```
select Meat {
 dishes := ...
}
filter exists .dishes
```
   
FULL OUTER JOIN
===============

These joins are so rare in practice that introducing some syntax syntactic sugar for them seemed premature.  We can always do that, of course, especially when lambdas are implemented, so it could be as simple as `select full_outer_join_table(Employer, Employee, condition := ((employer, employee) => employee.employer = employer))` (the syntax is totally made up here, the point is to show that rarely used patterns can be replaced with standard library once you can pass expressions as function arguments).

**Reply by elprans** [reply]
> If you have collections that only work in an "outer" fashion, I suppose you can add a semi-join to exclude the persons without children at all, but that's a bit less efficient as you join on the children twice:

That's true currently, however one great thing about EdgeQL is that we can teach the compiler to optimize joins, so the below

```
select Person {
  max_degree := max(.<parent[is Child].degree)
}
filter
  exists .<parent[is Child]
```

can be a single inner join.

**Reply by jods4** [reply]
Thanks for the infos!

I totally concur with (1.), explicit joins are very often unrequired noise, esp. if your model already defines the relationship. This is already apparent in many ORM that just let you traverse collections on your object, in a fashion that's reminiscent of EdgeDB.

As well as with (3.), nested structures are what you want/need 90% of the time and is much easier to work with than tuples.

Optimization-wise, it'd indeed be nice if `filter exists` could notice the that the relationship is joined and turn it into an INNER JOIN. 
Another interesting shortcut can be: in cases where the model has a non-optional 1..* relationship (not sure if this actually captured by EdgeDB modelling or not), it could use INNER instead of OUTER automatically.

The syntax you have in the product already is nice; I think the _full outer_ variant might need a little more creativity if you ever go that way.
Sure, it's not the most frequently used tool, but when you need it and you don't have it... it's a bit of a pain.

That's actually an interesting point because it's a risk I'm evaluating with new query languages like this. How much features do they support, will I ever encounter shortcomings? Here we write complex queries that use analytical functions (lead, lag, running sums, moving averages, partitioned counts), recursive CTEs, full joins, full text search, custom DB functions (such as encryption primitives, bitfield manipulations), we stream large blobs, and more. 
Adopting new tools like EdgeDB is a little scary. If there was an escape hatch where you can drop down to raw SQL in unsupported edge cases, that would be slightly reassuring.

> We are working on a more elaborate "EdgeQL for SQL users", but this discussion is a great opportunity to summarize that.

It's great to know that "EdgeQL for SQL users" is on your roadmap! 👀

I believe there are two broad families of users:
- Those (most?) who don't look too much into how their SQL DB works, maybe they even do everything with an ORM and code migrations already. They don't have stringent requirements on the RDBMS side. For them EdgeDB would be a perfect fit, and there's no need for more documentation.
- Those who perform crazy SQL workloads in fine-tuned DB schemas. I personally am in this bucket (we run _crazy_ queries, with amazing perf). I am still very interested in EdgeDB because like you I can see the shortcomings of the current way. We could be a lot more productive. But I'm very cautious about _how_ EdgeDB performs its magic, what the queries will look like, if I'll be able to fine-tune, etc.

From my experience, the latter is a minority, that maybe could even be ignored from a business perspective 😞

**Reply by elprans** [reply]
> model has a non-optional 1..* relationship (not sure if this actually captured by EdgeDB modelling or not)

It is: `required multi link -> Other`

> it could use INNER instead of OUTER automatically

EdgeDB actually always uses `INNER` SQL joins (or equivalent `WHERE` conditions).  The "left"-ness we discussed above is logical, not necessarily how it's translated to SQL.

> Here we write complex queries that use analytical functions (lead, lag, running sums, moving averages, partitioned counts), recursive CTEs, full joins, full text search.

Understandable.  Our goal for EdgeQL is for it to be a "better SQL", including in cases like these.  We're not there yet, but the gap is closing quickly.  Windowing and transitive closure are on the near term implementation TODO list.

> If there was an escape hatch where you can drop down to raw SQL in unsupported edge cases, that would be slightly reassuring.

Our position is that providing an easy "escape hatch" is detrimental in the long term, because it removes the pressure on the abstraction (EdgeQL in this case) to actually solve the use case appropriately.

> From my experience, the latter is a minority, that maybe could even be ignored from a business perspective

I think we can successfully satisfy both sides of the the spectrum :-)  In the end, when you write SQL queries, you are working with an abstract query planner and use `explain` to understand and tune the queries.  This wouldn't be different with EdgeDB once `explain` lands (the plan is to translate PostgreSQL `explain` output).
